api/src/opentrons/protocol_api/core/engine/module_core.py

```
@@ -739,6 +809,127 @@ def empty(self, message: str) -> None:
+        """Get the amount of space currently available for labware."""
+        max_lw = self.get_max_storable_labware()
+        current = len(
+            self._engine_client.state.modules.stacker_contained_labware(self._module_id)
+        )
+        return max_lw - current
+
+    def _predict_storable_count(self, labwares: _CoreTrio) -> int:
+        definitions = [
+            x.get_engine_definition()
+            for x in [labwares.adapter, labwares.primary, labwares.lid]
+            if x is not None
+        ]
+        pool_height = self._engine_client.state.geometry.get_height_of_labware_stack(
+            definitions
+        )
+        overlap = self._engine_client.state.labware.get_labware_overlap_offsets(
+            definitions[-1], definitions[0].parameters.loadName
+        )
+        return self._engine_client.state.modules.stacker_max_pool_count_by_height(
+            self._module_id, pool_height, overlap.z
+        )
+
+    def get_max_storable_labware_from_list(
+        self,
+        labware: Sequence[LabwareCore],
+    ) -> Sequence[LabwareCore]:
+        """Limit the passed list to how many labware can fit in a stacker."""
+        if not labware:
+            return labware
+        max_count = self._predict_storable_count(
+            self._core_groups_from_primary_core(labware[0])
+        )
+        return labware[:max_count]
+
+    def get_current_storable_labware_from_list(
+        self,
+        labware: Sequence[LabwareCore],
+    ) -> Sequence[LabwareCore]:
+        """Limit the passed list to how many labware can fit in the stacker right now."""
+        if not labware:
+            return labware
+        max_count = self._predict_storable_count(
+            self._core_groups_from_primary_core(labware[0])
+        )
+        current = len(
+            self._engine_client.state.modules.stacker_contained_labware(self._module_id)
+        )
+        total = max_count - current
+        return labware[:total]
```

What happens to the labware objects that don't get returned? Assuming a user can pass in a list of labware with some predefined states, does that mean those labware states could be silently lost? 

We want to be able to pass in a larger list than what is storable; should we either (1) enforce the list of objects cannot have predefined states, or (2) return the list of un-storable labware from the list so it can be loaded later or somehow store them in stacker memory so they can be used to fill the next time?

---

well, they're in there still, because for api `Labware` objects to exist they have to have been loaded to the engine, so the states aren't really lost. I think we'd want the documents to say "hey don't do `labware = stacker.get_max_storable_labware_from_list(labware)`, save it in a new variable or put it in the call".

(2) is kind of an interesting idea where you make it a protocol api local thing like instrument tipracks used to be, that's kind of cool